### PR TITLE
Adding Conv2d_bwd_data tuning parameters translation

### PR DESCRIPTION
**Pull request by jerryyin**
* Added populateDimVal and populateSeqVal to parse lowering data
  structures
* Added obtain*GemmVectorizable/VecLen for determining vectorization
  parameters
* Added bwd_data specific parameters related with atomicadd

### CODE REVIEW

**Comment by whchung** [comment]
@asroy
@zjing14

File: mlir/lib/Dialect/MIOpenOps/CppOutput/gridwise_convolution_implicit_gemm_v4r4.cpp [hunk 1]
```diff
@@ -702,91 +700,187 @@ void ObtainModuleInfo(ModuleOp &m,
...
+    // K's position is 2, vectorization legnth is last dimension
+    if (dimIndexVal["c"].first == 3) {
+      vecLen = dimIndexVal["c"].second;
+    } else if (dimIndexVal["y"].first == 3) {
+      vecLen = dimIndexVal["y"].second;
+    } else {
+      vecLen = dimIndexVal["x"].second;
+    }
+  }
+}
```
Comment (jerryyin) [inline mlir/lib/Dialect/MIOpenOps/CppOutput/gridwise_convolution_implicit_gemm_v4r4.cpp:791]: @asroy Could you do a sanity check for the filter `kcyx` layout vectorization length?

FYI `dimIndexVal` is the a String to pair mapping which holds the 
 - key: string representation of dimension, for example, k
 - value pair: first: index of the  dimension, for example, 0 for k in kcyx. 
                       second: value of the dimension, for example, 128 for number of filters.

File: mlir/lib/Dialect/MIOpenOps/CppOutput/gridwise_convolution_implicit_gemm_v4r4.cpp [hunk 2]
```diff
@@ -33,8 +33,6 @@ std::string resultStr;
 
 class TunableParameters : public TunableParametersBase {
 public:
-  TunableParameters() : TunableParametersBase("gridwise_convolution_implicit_gemm_v4r4.yaml") {}
-
   virtual void customInit() override {
     // parameters truly tunable.
     params["CK_PARAM_TUNABLE_GEMM_M_PER_BLOCK"] = 128;
```
Comment (jerryyin) [inline mlir/lib/Dialect/MIOpenOps/CppOutput/gridwise_convolution_implicit_gemm_v4r4.cpp:38]: I have walked through those parameters with @asroy in a remote session and make sure those parameters can work for both forward and backward_data. But just to make sure, could you take a second look to sanity check this as well?

**Comment by zjing14** [comment]
@jerryyin More changes than I thought. Looks like the majority of changes are related to the vector length logic. Do we need a unified function of vector length calculation for all three directions?

**Comment by jerryyin** [comment]
@zjing14 Right, I take the chance to do some refactoring in this PR too. By breaking the vector length calculation I managed to share a decent amount of code path between forward and backward data path. There will be a lot more duplication if I continue down the original path.

Regarding the suggestion to unify all three directions, I haven't find there's a strong need in this. It is because the code as is now popuplates the cflags one section at a time. I find it easier to handle each separately. Please feel free to comment if you think it can be better unified in a particular way.

**Pull request merged**